test/solvers/functional/functional_test.py

```diff
@@ -0,0 +1,602 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Test for the Functional class."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+# External
+import numpy as np
+import pytest
+
+# Internal
+import odl
+from odl.util.testutils import all_almost_equal, almost_equal, noise_element
+
+# Places for the accepted error when comparing results
+PLACES = 8
+
+
+def test_derivative():
+    """Test for the derivative of a functional.
+
+    The test checks that the directional derivative in a point is the same as
+    the inner product of the gradient and the direction, if the gradient is
+    defined.
+    """
+
+    # Discretization parameters
+    n = 3
+
+    # Discretized spaces
+    space = odl.uniform_discr([0, 0], [1, 1], [n, n])
+
+    x = space.element(np.random.standard_normal((n, n)))
+
+    y = space.element(np.random.standard_normal((n, n)))
+    epsK = 1e-8
+
+    F = odl.solvers.functional.L2Norm(space)
```

Comment:
Perhaps create a fixture of functionals and verify that this works for all of them? You can catch the `NotImplemented` case and consider it a success as well.